### PR TITLE
HDDS-11908. Snapshot diff DAG traversal should not skip node based on prefix presence

### PR DESCRIPTION
**Pull request by swamirishi**
## What changes were proposed in this pull request?
Snapshot diff DAG traversal should not skip node based on prefix presence. 

Consider the case 

Snap1 has 1, 2, 3 files

Say 1,4 compacted to => 5

And 5, 6 => 7

Snap 2 has 7, 2, 3 file.

Now diff b/w Snap1 & Snap2 would be in 4 & 6 sst files.

Consider 1 had +k & 4 had -k

Now 5 would have neither of these files.

But as per this condition:
https://github.com/apache/ozone/blob/b11b80707d4e6721c06966936ad87063ec107da6/hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java#L1009-L1016 traversal on would skip node 5 thus traversal would never reach 1,4. We should not have this optmization check. This would end up not using the compaction even though it is not there.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11908

## How was this patch tested?
Unit tests

### CODE REVIEW

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java
Comment (hemantk-12) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:118]: Remove this visibility tag.

File: hadoop-hdds/rocksdb-checkpoint-differ/src/test/java/org/apache/ozone/rocksdiff/TestRocksDBCheckpointDiffer.java
Comment (hemantk-12) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/test/java/org/apache/ozone/rocksdiff/TestRocksDBCheckpointDiffer.java:1797]: Tests (`testShouldSkipNode` and `testShouldSkipNodeEdgeCase`) should be moved to `TestRocksDiffUtils`. If not part of this PR, create a follow-up task.

**Review by prashantpogde — APPROVED** [review]
LGTM. Thanks @swamirishi for making this change.

**Review by hemantk-12 — APPROVED** [review]
Thanks for the patch @swamirishi.

LGTM+1.

Wait for the green CI before the merge and create [a follow-up task](https://github.com/apache/ozone/pull/7567/files#r1881239933).

**Comment by swamirishi** [comment]
> > ## How was this patch tested?
> > Unit tests
> 
> Sorry, I don't think that's correct. The only changed test is for calling `shouldSkipNode` from its new home, `RocksDiffUtils`. The behavior change in `RocksDBCheckpointDiffer` did not require changing any tests, so this case does not seem to be covered by tests.
> 
> Maybe "tested manually" would be more appropriate.

I have added unit test for this case as part of this PR
https://github.com/apache/ozone/pull/7563/commits/8548a539bd85c77c58a9ef0f5da55ce1f43e6309

**Comment by swamirishi** [comment]
Thanks for the reviewing the patch @hemantk-12 @prashantpogde @adoroszlai

**Comment by adoroszlai** [comment]
> I have added unit test for this case as part of this PR [8548a53](https://github.com/apache/ozone/commit/8548a539bd85c77c58a9ef0f5da55ce1f43e6309)

Which PR?

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java
Comment (swamirishi) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:118]: it is visible for testing since this shouldSkipNode function is being used only by tests and a function inside RocksDiffUtils. We could have made it private if not for the tests.

**Pull request merged**
